### ISSUE TITLE
Expand the object limit to over 32,000 instances

### ISSUE THREAD

**Issue by Morgerion** [defold/defold#9302]
It would be great to increase the limit on the number of instances to 100K.
32K is very low for medium/large projects.

To make it clear how small 32K is - here are some real examples i have already encountered:

1. To form a mini-map of a level, you need to render a field of 200x200 cells, and each cell can contain several GO-objects (the object itself + the shadow under it + the underlying cell). In this way you can easily get more than 32K objects.

2. Bullet-hell game. Each projectile is accompanied by shards, hitting enemies generates shards, at high levels these objects become very many. There can be more than 32K objects on the screen, whose behaviour is calculated individually through native expansion. This game works fine on PC, but the 32K limit is sometimes suddenly reached at the most intense moment.

3. Songs of Conquest-like game. In each cell of the playing field there can be several GO-objects (the object itself + the shadow under it + the underlying cell + debug-info). There are usually 40x40 cells on the screen, so 40*40*4 = 6400 sprites. At the same time it is possible to zoom the camera, which increases the area of visibility in 2 times, while the area increases in 4 times, totalling 6400*4 = 25K. The limit is not reached, but it's pretty close. Any small change in configuration could hit it.

Technically, it is possible to circumvent this restriction. For example, divide the game content into several collections. But it is technically possible not always - sometimes it is possible through ridiculous and unnecessary complication of architecture, which in fact will solve only one problem (overcoming the limit) but will create other problems.

The most preferable solution is to simply increase the limit restriction. Not everyone will obviously reach such high values, but those who need it will be very happy.

**Comment by JCash** [comment]
I think this is an acceptable change.
The game object instances would grow by 14 bytes each, which shouldn't be too noticable for smaller games.

A more curious implementation detail that needs checking, is why the current max instances is:
```
// Invalid instance index. Implies that maximum number of instances is 32766 (ie 0x7fff - 1)
const uint32_t INVALID_INSTANCE_INDEX = 0x7fff;
```

E.g. are we really relying on a bit mask here, or was that just some premature design?
In the gameobject.cpp, at a quick glance, I couldn't find and index operations involving masks, and the instance indices are private to that system.
Which suggests that it should be a fairly easy fix.

**Comment by AGulev** [comment]
> E.g. are we really relying on a bit mask here, or was that just some premature design?

I found this one 

https://github.com/defold/defold/blob/3ede22e06920e3b6e2aa30f1d1170b6f9f597c99/engine/gameobject/src/gameobject/gameobject_profile.cpp#L70-L73

**Comment by JCash** [comment]
Good find. Still a private implementation which would fit within the 64 bit value there, so should also be an easy update.

**Comment by akhleung** [comment]
Just want to mention that I'm working on a sandbox / builder style of game, so a higher limit would be nice in order to allow players to create larger environments.

**Comment by AGulev** [comment]
Some rough estimations.

# Why it’s limited (~32k)

  - 15‑bit indices: The instance index and several related indices are stored in 16‑bit bitfields with 15 bits reserved for
  the index and 1 bit for flags. See:

     https://github.com/defold/defold/blob/5eb6396edeff13472fbb2250137942114595ec43/engine/gameobject/src/gameobject/gameobject_private.h#L156-L158
https://github.com/defold/defold/blob/5eb6396edeff13472fbb2250137942114595ec43/engine/gameobject/src/gameobject/gameobject_private.h#L173-L174


# Possible solutions

##  Changes Required for full 16 bit indices (~65k instances)

The simplest fix with minimal impact would be to use all 16-bit for the index and move flags into the padding in the structure we already have:

https://github.com/defold/defold/blob/5eb6396edeff13472fbb2250137942114595ec43/engine/gameobject/src/gameobject/gameobject_private.h#L126-L181

It will require the following changes:

  - engine/gameobject/src/gameobject/gameobject_private.h
      - Set INVALID_INSTANCE_INDEX to 0xffff.
      - Repack Instance bitfields so indices are full 16-bit, and flags move to the existing 16-bit flag word:
          - Keep the first flag word but use the pad bits:
              - Before: m_Depth:8, m_Initialized:1, m_Bone:1, m_Generated:1, m_Pad:5
              - After:  m_Depth:8, m_Initialized:1, m_Bone:1, m_Generated:1, m_ToBeDeleted:1, m_ToBeAdded:1, m_Pad:3
          - Widen index fields from 15→16 bits and drop the 1-bit flag/pad that followed each:
              - m_Index : 16     (remove the old m_ToBeDeleted : 1 here)
              - m_LevelIndex : 16 (remove the old m_Pad2 : 1)
              - m_SiblingIndex : 16 (remove the old m_ToBeAdded : 1)
              - m_FirstChildIndex : 16 (remove the old m_Pad4 : 1)
          - Keep m_Parent : 16, m_NextToDelete : 16, m_NextToAdd as uint16_t (unchanged).
          - Constructor defaults that assign INVALID_INSTANCE_INDEX continue to work with 0xffff.
  - engine/gameobject/src/gameobject/gameobject.cpp
      - NewCollection limit: keep condition if (max_instances > INVALID_INSTANCE_INDEX); update the error string to reflect
  the new limit (show INVALID_INSTANCE_INDEX).
      - Default capacity setter: make the range check consistent with the above (currently uses >= INVALID_INSTANCE_INDEX
          - Change to if (capacity > INVALID_INSTANCE_INDEX || capacity == 0) return RESULT_INVALID_OPERATION;
  - com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties
      - Raise collection.max_instances.maximum from 32765 to 65535.
  - engine/gameobject/src/gameobject/gameobject.h
      - Update doc comments mentioning valid ranges (e.g. SetCollectionDefaultCapacity) from “0–32766” to “0–65535”.
  - No other refactors are needed:
      - dmIndexPool16, uint16_t level indices, and traversal logic all remain valid with a 16-bit sentinel of 0xffff.
      - gameobject_profile.cpp uses component_count_offset = 0xFFFF and checks still hold; no change required.

###  Memory Impact

  - At the same configured collection.max_instances: effectively none from the struct layout, if you repack flags as above
  (indices still occupy 16 bits total and flags move into existing pad).

##  Changes Required for 32‑bit indices

This approach would require significantly more changes and would be more impactful in terms of memory usage.

- Core constants and types
      - engine/gameobject/src/gameobject/gameobject_private.h
          - Set INVALID_INSTANCE_INDEX to 0xffffffff (new sentinel).
          - Widen all instance index fields to 32‑bit:
              - m_Parent → uint32_t
              - m_Index → uint32_t (remove per‑field 1‑bit flag)
              - m_LevelIndex → uint32_t
              - m_NextToDelete → uint32_t
              - m_NextToAdd → uint32_t
              - m_SiblingIndex → uint32_t
              - m_FirstChildIndex → uint32_t
          - Move m_ToBeDeleted/m_ToBeAdded into the existing 16‑bit flag word (with m_Depth, m_Initialized, m_Bone,
  m_Generated), reusing pad bits so no new words are added.
          - Collection: change m_InstanceIndices from dmIndexPool16 to dmIndexPool32.
          - Collection: change m_LevelIndices[MAX_HIERARCHICAL_DEPTH] from dmArray<uint16_t> to dmArray<uint32_t>.
  - Engine logic using indices
      - engine/gameobject/src/gameobject/gameobject.cpp
          - Update types/locals to 32‑bit where they hold instance indices:
              - uint16_t instance_index = collection->m_InstanceIndices.Pop(); → uint32_t
              - uint16_t level_index = instance->m_LevelIndex; and similar → uint32_t
              - Functions EraseSwapLevelIndex, InsertInstanceInLevelIndex, ExpandLevel: dmArray<uint16_t>& →
  dmArray<uint32_t>&, all internal variables to uint32_t.
              - Loops/fields using m_InstancesTo(Add|Delete)(Head|Tail) now operate on uint32_t.
          - NewCollection/default capacity range checks:
              - if (max_instances >= INVALID_INSTANCE_INDEX) { … }
              - if (capacity >= INVALID_INSTANCE_INDEX || capacity == 0) return RESULT_INVALID_OPERATION;
      - engine/gameobject/src/gameobject/gameobject_profile.cpp
          - Disambiguation between gameobject indices and component iteration must change (the old “component_count_offset =
  0xFFFF” can no longer be ≥ invalid index).
          - Use:
              - const uint32_t invalid_index = 0xFFFFFFFF;
              - const uint32_t component_count_offset = invalid_index - 1;
              - If it->m_NextChild.m_Node == invalid_index, set it to component_count_offset to start component iteration.
              - Classify:
                  - gameobject when index < component_count_offset
                  - component when index == component_count_offset
              - Remove/replace the old DM_STATIC_ASSERT(component_count_offset >= INVALID_INSTANCE_INDEX, …) with logic
  consistent with the above.
          - Also update the root level type: const dmArray<uint32_t>& root_level = collection->m_LevelIndices[0];
      - engine/gameobject/src/gameobject/res_collection.cpp
          - Resource size accounting:
              - size += collection->m_InstanceIndices.Capacity()*sizeof(uint32_t); (was sizeof(uint16_t)).
  - Tooling
      - com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties
          - Raise [collection] max_instances.maximum to a sane upper bound for UI (Java int safe). Suggest 2147483646. The
  engine allows up to 0xFFFFFFFF - 1, but the UI should avoid absurd values.
  - Docs and ranges
      - engine/gameobject/src/gameobject/gameobject.h
          - Update comments for valid ranges (e.g., capacity “0– 2147483646 ”).

  Memory Impact (vs current 16‑bit design)

  - Per reserved slot (capacity)
      - Instance index pool: +2 bytes/slot (32‑bit vs 16‑bit).
  - Per live instance (allocated)
      - Instance struct: +14–16 bytes (all 16‑bit index fields widened to 32‑bit; small extra padding possible).
      - Level index storage: +2 bytes/instance total across all levels (each instance appears in exactly one level array).
  - Per collection (heads/tails)
      - m_InstancesTo* fields widened: +8 bytes total (negligible).
  - Example deltas
      - Keeping capacity (no change to max_instances):
          - No capacity growth cost; you still pay per live instance: ~+16–18 bytes each.
      - Increasing capacity by N
          - Index pool reserved: +2·N bytes.
          - Live instances cost: +16–18 bytes each as they are created.
  - Context for large capacities
      - Existing big arrays unchanged: m_WorldTransforms ~64 bytes/slot; m_Instances pointer array ~8 bytes/slot;
  m_IDToInstance ~32 bytes/entry. Those dominate at large max_instances. Moving to 32‑bit indices doesn’t change these, but
  if you scale capacity significantly, total reserved memory grows accordingly.

A very rough estimate is ~200–250 kb of additional memory per 10,000 game objects (not that critical because of memory itself, but worse data locality).

  Caveats

  - This is a broad change touching core data layouts; compile errors will surface remaining uint16_t index assumptions
  (tests/profiling/iteration code).
  - The profiling iterator’s range logic must be updated as above; it’s the one spot where the old “< INVALID_INSTANCE_INDEX”
  shortcut breaks.

----

I recommend choosing the full `16 bit option`, as it requires far less work and has virtually no impact on memory or performance and gives 65 534 objects limit. We can then see if we need more in practice in the future.

**Comment by AGulev** [comment]
We’re raising the limit to 65,534 for now - this is the maximum achievable without major rework or performance penalties.  
Although this is lower than what you requested, I’ll close this issue. If you later find that the limit is still insufficient for your project, please feel free to open a new issue.
